### PR TITLE
Wl kernel

### PR DESCRIPTION
**Pull request by KaterynaMelnyk**
WL implementation

### CODE REVIEW

File: graph.h [hunk 1]
```diff
@@ -32,9 +33,7 @@ namespace Graph {
 
     template <typename TYPE_WEIGHT, typename TYPE_NODE_LABEL>
     Graph<TYPE_WEIGHT, TYPE_NODE_LABEL>::Graph(size_t numOfNodes) : numberOfNodes(numOfNodes) {
-        for (size_t i = 0; i < numberOfNodes; ++i) {
-            nodeLabels[i] = 0;
-        }
+        std::vector<TYPE_NODE_LABEL> nodeLabels(numberOfNodes, 0);
```
Comment (SunshineJohn) [inline graph.h]: What are you doing here? It's a local variable which will be destroyed after leaving the constructor.

File: graph.h [hunk 2]
```diff
@@ -43,6 +42,10 @@ namespace Graph {
     }
 
     template <typename TYPE_WEIGHT, typename TYPE_NODE_LABEL>
+    std::vector<TYPE_NODE_LABEL> Graph<TYPE_WEIGHT, TYPE_NODE_LABEL>::GetAllNodeLabels() const{
```
Comment (SunshineJohn) [inline graph.h]: `const {`

File: graph.h [hunk 3]
```diff
@@ -71,6 +74,18 @@ namespace Graph {
         return adjList.at(indexOfNode);
     }
 
+    template <typename TYPE_WEIGHT, typename TYPE_NODE_LABEL>
+    std::vector<int> Graph<TYPE_WEIGHT, TYPE_NODE_LABEL>::GetNodeNeighbors(size_t indexOfNode) {
```
Comment (SunshineJohn) [inline graph.h]: std::vector<size_t> is better, I think.

File: graph.h [hunk 4]
```diff
@@ -71,6 +74,18 @@ namespace Graph {
...
+        std::list<std::pair<size_t, TYPE_WEIGHT>> currentAdjList = GetAdjList(indexOfNode);
```
Comment (SunshineJohn) [inline graph.h]: `const std::list<std::pair<size_t, TYPE_WEIGHT>>&`

File: graph_kernels.h [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+//TODO: implementation of GraphletKernel
+//TODO: implementation of the DeepWalk
+
+namespace Graph {
+    template <typename TYPE_NODE_LABEL>
+    std::vector<TYPE_NODE_LABEL> GetNodeLabelsFromVector(std::vector<int> nodeIndeces, std::vector<TYPE_NODE_LABEL> allNodeLabels) {
+        std::vector<TYPE_NODE_LABEL> nodeNeighborsLabels;
+
+        for (std::vector<int>::iterator it = nodeIndeces.begin(); it != nodeIndeces.end(); ++it) {
```
Comment (SunshineJohn) [inline graph_kernels.h]: Why not `auto it`?

File: graph_kernels.h [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+#include "graph.h"
+#include "utils.h"
+
+//TODO: implementation of GraphletKernel
+//TODO: implementation of the DeepWalk
+
+namespace Graph {
+    template <typename TYPE_NODE_LABEL>
+    std::vector<TYPE_NODE_LABEL> GetNodeLabelsFromVector(std::vector<int> nodeIndeces, std::vector<TYPE_NODE_LABEL> allNodeLabels) {
```
Comment (SunshineJohn) [inline graph_kernels.h]: `nodeIndices`

File: graph_kernels.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        std::vector<int> countUniqueLabels;
+        typename std::vector<TYPE_NODE_LABEL>::iterator it;
+
+        for (size_t i = 0; i < uniqueNodeLabels.size(); ++i) {
+            int currectCount = std::count(allNodeLabels.begin(), allNodeLabels.end(), uniqueNodeLabels[i]);
+
+            countUniqueLabels.push_back(currectCount);
+        }
+        return countUniqueLabels;
+
```
Comment (SunshineJohn) [inline graph_kernels.h]: Remove empty line.

File: graph_kernels.h [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        for (std::vector<int>::iterator it = nodeIndeces.begin(); it != nodeIndeces.end(); ++it) {
+            nodeNeighborsLabels.push_back(allNodeLabels[*it]);
+        }
+        return nodeNeighborsLabels;
+    }
+
+    template <typename TYPE_NODE_LABEL>
+    std::vector<int> CountLabelsInGraph(std::vector<TYPE_NODE_LABEL> allNodeLabels, std::vector<TYPE_NODE_LABEL> uniqueNodeLabels) {
+        std::vector<int> countUniqueLabels;
+        typename std::vector<TYPE_NODE_LABEL>::iterator it;
```
Comment (SunshineJohn) [inline graph_kernels.h]: Do you still need it?

File: graph_kernels.h [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+            countUniqueLabels.push_back(currectCount);
+        }
+        return countUniqueLabels;
+
+    }
+
+    template <typename TYPE_WEIGHT, typename TYPE_NODE_LABEL>
+    std::vector<std::vector<float>> WLSubtreeKrenel(const std::vector<Graph<TYPE_WEIGHT, TYPE_NODE_LABEL>>& vectorOfGraph, size_t numberOfIteration) {
+        size_t numberOfGraphs = vectorOfGraph.size();
```
Comment (SunshineJohn) [inline graph_kernels.h]: `const size_t`

File: graph_kernels.h [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        std::vector<std::vector<int>> kernelMatrix(numberOfGraphs, std::vector<int>(numberOfGraphs, 0));
+        std::vector<TYPE_NODE_LABEL> allNodeLabels;
+        std::vector<int> allCompressedNodeLabels;
+
+        std::vector<bool> comprassedGraph(numberOfGraphs, false);
+
+        std::vector<std::vector<int>> newGraphs;
+
+        for (auto graph : vectorOfGraph) {
```
Comment (SunshineJohn) [inline graph_kernels.h]: `auto&`

File: graph_kernels.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+            std::vector<int> newNodeLabels;
+
+            std::vector<TYPE_NODE_LABEL> nodeLabels = graph.GetAllNodeLabels();
```
Comment (SunshineJohn) [inline graph_kernels.h]: You can do like `allNodeLabels.insert(allNodeLabels.end(), nodeLabels.begin(), nodeLabels.end())`

File: graph_kernels.h [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        for (size_t i = 0; i < numberOfGraphs; ++i) {
+            std::vector<TYPE_NODE_LABEL> currentNodeLabels = vectorOfGraph[i].GetAllNodeLabels();
+            currentNodeLabels.insert(currentNodeLabels.end(), newGraphs[i].begin(), newGraphs[i].end());
+
+            countLabels.push_back(CountLabelsInGraph(currentNodeLabels, allNodeLabels));
+        }
+
+        for (size_t i = 0; i < countLabels.size(); ++i) {
+            for (size_t j = i; j < countLabels.size(); ++j) {
+                int dotProduct = std::inner_product(countLabels[i].begin(), countLabels[i].end(), countLabels[j].begin(), 0);
```
Comment (SunshineJohn) [inline graph_kernels.h]: `const int`

File: graph_kernels.h [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+                kernelMatrix[i][j] = dotProduct;
+                kernelMatrix[j][i] = dotProduct;
+            }
+        }
+
```
Comment (SunshineJohn) [inline graph_kernels.h]: Remove empty line.

File: graph_kernels.h [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+            newGraphs.push_back(newNodeLabels);
+        }
+
+        UniqueValuesInVector(allNodeLabels);
+        UniqueValuesInVector(allCompressedNodeLabels);
+        allNodeLabels.insert(allNodeLabels.end(), allCompressedNodeLabels.begin(), allCompressedNodeLabels.end());
+
+        std::vector<std::vector<int>> countLabels;
+        for (size_t i = 0; i < numberOfGraphs; ++i) {
+            std::vector<TYPE_NODE_LABEL> currentNodeLabels = vectorOfGraph[i].GetAllNodeLabels();
```
Comment (SunshineJohn) [inline graph_kernels.h]: `const std::vector<TYPE_NODE_LABEL>`

File: graph.h (hunk 1 above)
Comment (KaterynaMelnyk) [reply graph.h]: Why? nodeLabels is a member of the class

File: graph_kernels.h (hunk 12 above)
Comment (KaterynaMelnyk) [reply graph_kernels.h]: Why?

File: DfsBfs.h [hunk 15]
```diff
@@ -78,9 +78,9 @@ namespace Trees {
     }
 
     template<typename TYPE>
-    std::vector<TYPE> SearchInRange(const BinaryTree<TYPE>& tree, int leftBorder, int rightBorder) {
+    std::vector<int> SearchInRange(const BinaryTree<TYPE>& tree, int leftBorder, int rightBorder) {
```
Comment (SunshineJohn) [inline DfsBfs.h:81]: `int`?

File: graph.h [hunk 16]
```diff
@@ -32,16 +33,18 @@ namespace Graph {
...
-        }
     }
 
     template <typename TYPE_WEIGHT, typename TYPE_NODE_LABEL>
     size_t Graph<TYPE_WEIGHT, TYPE_NODE_LABEL>::GetNumberOfNodes() const {
         return numberOfNodes;
     }
 
+    template <typename TYPE_WEIGHT, typename TYPE_NODE_LABEL>
+    std::vector<TYPE_NODE_LABEL> Graph<TYPE_WEIGHT, TYPE_NODE_LABEL>::GetAllNodeLabels() const{
```
[the remark this answers is not in the document]
Comment (KaterynaMelnyk) [reply graph.h:44]: Why do we need & ?

File: graph.h [hunk 17]
```diff
@@ -83,7 +98,7 @@ namespace Graph {
     }
 
     template <typename TYPE_WEIGHT, typename TYPE_NODE_LABEL>
-    void Graph<TYPE_WEIGHT, TYPE_NODE_LABEL>::AddNodeLabels(std::vector<TYPE_NODE_LABEL> vectorOfNodeLabels) const{
+    void Graph<TYPE_WEIGHT, TYPE_NODE_LABEL>::AddNodeLabels(std::vector<TYPE_NODE_LABEL>& vectorOfNodeLabels) {
```
Comment (SunshineJohn) [inline graph.h:101]: `const std::vector<TYPE_NODE_LABEL>&`

**Pull request merged**